File: backends/node/services/geo/geo-client/src/handlers/x/find-whois.ts
import { BaseHandler, type IHandlerContext, validators } from "@d2/handler";
import { D2Result } from "@d2/result";
import { CircuitBreaker, CircuitState } from "@d2/utilities";
import type { WhoIsDTO, FindWhoIsRequest, FindWhoIsResponse, GeoServiceClient } from "@d2/protos";
import type { MemoryCacheStore } from "@d2/cache-memory";
import { Metadata } from "@grpc/grpc-js";
import { z } from "zod";
import type { GeoClientOptions } from "../../geo-client-options.js";
import { GEO_CACHE_KEYS } from "../../cache-keys.js";
import { Complex } from "../../interfaces/index.js";

type Input = Complex.FindWhoIsInput;
type Output = Complex.FindWhoIsOutput;

/**
 * Handler for finding WhoIs data by IP address and fingerprint.
 * Checks local LRU memory cache first, falls back to Geo service gRPC.
 * Fail-open: never returns error results, always Ok with nullable WhoIsDTO.
 *
 * Mirrors D2.Geo.Client.CQRS.Handlers.X.FindWhoIs in .NET.
 */
export class FindWhoIs extends BaseHandler<Input, Output> implements Complex.IFindWhoIsHandler {
  override get redaction() {
    return Complex.FIND_WHOIS_REDACTION;
  }

  private readonly store: MemoryCacheStore;
  private readonly geoClient: GeoServiceClient;
  private readonly options: GeoClientOptions;
  private readonly circuitBreaker: CircuitBreaker;

  constructor(
    store: MemoryCacheStore,
    geoClient: GeoServiceClient,
    options: GeoClientOptions,
    context: IHandlerContext,
  ) {
    super(context);
    this.store = store;
    this.geoClient = geoClient;
    this.options = options;
    this.circuitBreaker = new CircuitBreaker({
      failureThreshold: options.circuitBreakerFailureThreshold,
      cooldownMs: options.circuitBreakerCooldownMs,
      onStateChange: (from, to) => {
        if (to === CircuitState.OPEN) {
          context.logger.warn(
            `Geo gRPC circuit breaker opened after ${options.circuitBreakerFailureThreshold} consecutive failures. ` +
            `Will probe in ${options.circuitBreakerCooldownMs}ms.`,
          );
        } else if (to === CircuitState.CLOSED && from === CircuitState.HALF_OPEN) {
          context.logger.info("Geo gRPC circuit breaker closed — service recovered.");
        }
      },
    });
  }

  private static readonly findWhoIsSchema = z.object({
    ipAddress: validators.zodIpAddress,
    fingerprint: z.string().min(1, "Fingerprint must not be empty"),
  }) as z.ZodType<Input>;

  protected async executeAsync(input: Input): Promise<D2Result<Output | undefined>> {
    // Validate input.
    const validation = this.validateInput(FindWhoIs.findWhoIsSchema, input);
    if (validation.failed) {
      return D2Result.bubbleFail(validation);
    }

    const cacheKey = GEO_CACHE_KEYS.whois(input.ipAddress, input.fingerprint);

    // Try cache first (null = negative cache sentinel, undefined = miss)
    const cached = this.store.get<WhoIsDTO | null>(cacheKey);
    if (cached !== undefined) {
      return D2Result.ok({ data: { whoIs: cached ?? undefined } });
    }

    // Cache miss — call Geo service (circuit breaker protects against sustained failures)
    let response: FindWhoIsResponse;
    try {
      const request: FindWhoIsRequest = {
        requests: [{ ipAddress: input.ipAddress, fingerprint: input.fingerprint }],
      };
      response = await this.circuitBreaker.execute(() =>
        new Promise<FindWhoIsResponse>((resolve, reject) => {
          this.geoClient.findWhoIs(
            request,
            new Metadata(),
            { deadline: Date.now() + this.options.grpcTimeoutMs },
            (err, res) => {
              if (err) reject(err);
              else resolve(res);
            },
          );
        }),
      );
    } catch {
      // Fail-open: handles gRPC errors AND circuit-open rejections identically.
      // When the circuit is open this returns instantly (no timeout wait).
      // Do not log input.ipAddress directly — it bypasses BaseHandler's redaction.
      if (this.circuitBreaker.state !== CircuitState.OPEN) {
        this.context.logger.warn(`gRPC call to Geo service failed. TraceId: ${this.traceId}`);
      }
      return D2Result.ok({ data: { whoIs: undefined } });
    }

    if (!response.result?.success || response.data.length === 0) {
      // Negative cache: avoid repeated gRPC calls for unknown IPs
      this.store.set(cacheKey, null, this.options.whoIsNegativeCacheExpirationMs);
      return D2Result.ok({ data: { whoIs: undefined } });
    }

    const whoIs = response.data[0]?.whois;

    // Cache the result
    if (whoIs !== undefined) {
      this.store.set(cacheKey, whoIs, this.options.whoIsCacheExpirationMs);
    }

    return D2Result.ok({ data: { whoIs } });
  }
}

export type { FindWhoIsInput, FindWhoIsOutput } from "../../interfaces/x/find-whois.js";

File: backends/dotnet/services/Geo/Geo.Client/CQRS/Handlers/X/FindWhoIs.cs
// -----------------------------------------------------------------------
// <copyright file="FindWhoIs.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace D2.Geo.Client.CQRS.Handlers.X;

using D2.Services.Protos.Geo.V1;
using D2.Shared.Handler;
using D2.Shared.Interfaces.Caching.InMemory.Handlers.R;
using D2.Shared.Interfaces.Caching.InMemory.Handlers.U;
using D2.Shared.Result;
using Grpc.Core;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using H = D2.Geo.Client.Interfaces.CQRS.Handlers.X.IComplex.IFindWhoIsHandler;
using I = D2.Geo.Client.Interfaces.CQRS.Handlers.X.IComplex.FindWhoIsInput;
using O = D2.Geo.Client.Interfaces.CQRS.Handlers.X.IComplex.FindWhoIsOutput;

/// <summary>
/// Handler for finding WhoIs data by IP address and user agent.
/// </summary>
/// <remarks>
/// This implementation checks the local memory cache first, then falls back to
/// the Geo service via gRPC if not cached. Results are cached for future lookups.
/// </remarks>
public class FindWhoIs : BaseHandler<FindWhoIs, I, O>, H
{
    private readonly IRead.IGetHandler<WhoIsDTO> r_cacheGet;
    private readonly IUpdate.ISetHandler<WhoIsDTO> r_cacheSet;
    private readonly GeoService.GeoServiceClient r_geoClient;
    private readonly GeoClientOptions r_options;

    /// <summary>
    /// Initializes a new instance of the <see cref="FindWhoIs"/> class.
    /// </summary>
    ///
    /// <param name="cacheGet">
    /// The cache get handler.
    /// </param>
    /// <param name="cacheSet">
    /// The cache set handler.
    /// </param>
    /// <param name="geoClient">
    /// The Geo service gRPC client.
    /// </param>
    /// <param name="options">
    /// The Geo client options.
    /// </param>
    /// <param name="context">
    /// The handler context.
    /// </param>
    public FindWhoIs(
        IRead.IGetHandler<WhoIsDTO> cacheGet,
        IUpdate.ISetHandler<WhoIsDTO> cacheSet,
        GeoService.GeoServiceClient geoClient,
        IOptions<GeoClientOptions> options,
        IHandlerContext context)
        : base(context)
    {
        r_cacheGet = cacheGet;
        r_cacheSet = cacheSet;
        r_geoClient = geoClient;
        r_options = options.Value;
    }

    /// <inheritdoc />
    protected override HandlerOptions DefaultOptions => new(LogOutput: false);

    /// <summary>
    /// Finds WhoIs data for the given IP address and user agent.
    /// </summary>
    ///
    /// <param name="input">
    /// The input containing IP address and user agent.
    /// </param>
    /// <param name="ct">
    /// The cancellation token.
    /// </param>
    ///
    /// <returns>
    /// A <see cref="ValueTask"/> containing a <see cref="D2Result{O}"/> with the WhoIs data
    /// if found; otherwise, null. Returns Ok with null on failures (fail-open behavior).
    /// </returns>
    protected override async ValueTask<D2Result<O?>> ExecuteAsync(
        I input,
        CancellationToken ct = default)
    {
        var cacheKey = $"whois:{input.IpAddress}:{input.UserAgent}";

        // Try cache first.
        var getR = await r_cacheGet.HandleAsync(new(cacheKey), ct);
        if (getR.CheckSuccess(out var cached) && cached?.Value is not null)
        {
            return D2Result<O?>.Ok(new O(cached.Value));
        }

        // Cache miss — call Geo service.
        FindWhoIsResponse response;
        try
        {
            response = await r_geoClient.FindWhoIsAsync(
                new FindWhoIsRequest
                {
                    Requests =
                    {
                        new FindWhoIsKeys
                        {
                            IpAddress = input.IpAddress,
                            Fingerprint = input.UserAgent,
                        },
                    },
                },
                cancellationToken: ct);
        }
        catch (RpcException ex)
        {
            // Fail-open: log warning and return null, not error.
            // Note: Do not log input.IpAddress as a scalar — it bypasses [RedactData]
            // destructuring. Use {@Input} for structured logging or omit PII entirely.
            Context.Logger.LogWarning(
                ex,
                "gRPC call to Geo service failed for {@Input}. TraceId: {TraceId}",
                input,
                TraceId);

            return D2Result<O?>.Ok(new O(null));
        }

        if (!response.Result.Success || response.Data.Count == 0)
        {
            // No data found or request failed — return null (fail-open).
            return D2Result<O?>.Ok(new O(null));
        }

        var whoIs = response.Data[0].Whois;

        // Store in cache.
        var setR = await r_cacheSet.HandleAsync(
            new(cacheKey, whoIs, r_options.WhoIsCacheExpiration),
            ct);

        if (setR.Failed)
        {
            // Note: Do not log input.IpAddress as a scalar — it bypasses [RedactData]
            // destructuring. Use {@Input} for structured logging or omit PII entirely.
            Context.Logger.LogWarning(
                "Failed to cache WhoIs for {@Input}. TraceId: {TraceId}",
                input,
                TraceId);
        }

        return D2Result<O?>.Ok(new O(whoIs));
    }
}
